### PR TITLE
[Integration Tests] Connection - 4 final tests.

### LINKED ISSUE #2390: [Integration Tests] Implement coverage for connection

### PR DESCRIPTION
**Pull request by ferdeen**
Fixes #2390.

Covers :-

- TS100_Connectivity_CanGetIPList
- TS102_Connectivity_CallConnect
- TS105_Connectivity_PreventConnectingToBannedNodes
- TS106_Connectivity_CanErrorHandleConnectionToNonExistingNodes

Note :-

- TS101_Connectivity_CallAddNode - Done in a previous PR.
- TS103_Connectivity_ListenDisabled - Ignored, as Listen Mode not implemented.
- TS104_Connectivity_IBD - Ignored, as Listen Mode not implemented.

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs [hunk 1]
```diff
@@ -23,8 +26,38 @@ public ConnectivityTests()
...
 
         [Fact]
-        public async Task WhenConnectingWithAddnodeConnectToPeerAndAnyPeersInTheAddressManager()
+        public void WhenStartingNodeWithAListOfIPAddressesTheAddressesArePropagatedToAnotherPeer()
         {
+            // TS100_Connectivity_CanGetIPList.
+
+            using (NodeBuilder builder = NodeBuilder.Create(this))
+            {
+                CoreNode node1 = builder.CreateStratisPosNode(this.Network).NotInIBD().Start();
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs]: Oh, please lowercase `this.Network` :)

File: src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs [hunk 2]
```diff
@@ -40,19 +73,123 @@ public async Task WhenConnectingWithAddnodeConnectToPeerAndAnyPeersInTheAddressM
...
+            sourceNode.FullNode.NodeService<IPeerAddressManager>().AddPeer(bannedNode.Endpoint, IPAddress.Loopback);
+
+            PeerAddress peerAddress = sourceNode.FullNode.NodeService<IPeerAddressManager>().Peers.FirstOrDefault();
+
+            if (peerAddress != null)
+                peerAddress.BanUntil = null;
+
+            return sourceNode;
+        }
+
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs]: Extra line

File: src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs [hunk 3]
```diff
@@ -23,8 +26,38 @@ public ConnectivityTests()
...
+                string hostname = string.Empty;
+
+                for (int i = 0; i < 10; i++)
+                {
+                    hostname = $"testnet{i}.stratisplatform.com";
+                    node2DNSSeeds.Add(new DNSSeedData(hostname, hostname));
+                }
+
+                node1DNSSeeds.Should().HaveSameCount(node2DNSSeeds);
```
Comment (zeptin) [inline src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs]: The name of this test seems to imply it is doing something that is not reflected in the test itself, unless I am mistaken.

File: src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs (hunk 1 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs]: Ok.

File: src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs (hunk 2 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs]: Hate extra lines :) thanks!

File: src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs (hunk 3 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Connectivity/ConnectivityTests.cs]: Ok.  Also addeding the correct implementation for this after @zeptin's comments.

_`getaddr` messages were sent from node 2 to node 1, and that node 1 sent a payload back to node 2 with peers in it (the `addr` payload)._

**Comment by ferdeen** [comment]
@fassadlr & @zeptin, 

I had difficulty trying to check that a peer address would propagate from one peer to the other.

`builder.CreateStratisPosNode` would build a node with `.SubstituteDateTimeProviderFor<MiningFeature>()`, setup.  This meant the UTCTime was set to  01/01/2018.

During `this.peerAddressManager.PeerSelector.SelectPeersForDiscovery()`, the method would check the peers `LastDiscoveredFrom` property.  The check  was to make sure the peer `LastDiscoveredFrom` date was less than 31/12/2017 (-24 hours from 01/01/2018).  So it was never picked up.  

I tried to use the custom node setup but still could not get the peer to propagate.

I didn't want to mock the logic out, as it would defeat creating an integration test.

However, I was able to write an alternative integration test.  Which would check that a  `AddressMananger`   peer connects to other nodes, and is available in `AddressManagerBehaviour`.

The test is : `EnsurePeerInAddressManagerAlsoConnectsAndSeenInPeerAddressManagerBehaviour`.

**Comment by fassadlr** [comment]
@ferdeen Ive updated your test so that we can ensure that peers being connected to, can be discovered from. More explanation in the test 👍 Once the build passes I will merge.

**Pull request merged**